File: src/main/GO/leetcode/lists/topinterview/LeetCode103.go
package topinterview

func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	res := make([][]int, 0)
	q := make([]*TreeNode, 0)
	flag := false
	q = append(q, root)
	for len(q) > 0 {
		size := len(q)
		level := make([]int, 0)
		for size > 0 {
			node := q[0]
			q = q[1:]
			level = append(level, node.Val)
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
			size--
		}
		if flag {
			Reverse(level, 0, len(level)-1)
		}
		flag = !flag
		res = append(res, level)
	}
	return res
}

File: src/main/java/leetcode/lists/topinterview/LeetCode103.java
package leetcode.lists.topinterview;

import common.*;

import java.util.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/23 17:52
 */
public class LeetCode103 {

    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        Queue<TreeNode> queue = new LinkedList<>();
        List<List<Integer>> list = new ArrayList<>();
        queue.add(root);
        boolean flag = false;
        while (!queue.isEmpty()) {
            //fix
            List<Integer> level = new ArrayList<>();
            int size = queue.size();
            while (size > 0) {
                TreeNode node = queue.poll();
                level.add(node.val);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
                size--;
            }
            if (flag) {
                //fix
                Collections.reverse(level);
            }
            list.add(level);
            flag = !flag;
        }
        return list;
    }
}
